librz/bin/format/mach0/dyldcache.c

```diff
@@ -1,295 +1,1076 @@
+		free(rebase_info);
+	}
+}
+
+static RzDyldLocSym *rz_dyld_locsym_new(RzBuffer *cache_buf, cache_hdr_t *hdr) {
+	if (!cache_buf || !hdr || !hdr->localSymbolsSize || !hdr->localSymbolsOffset) {
 		return NULL;
 	}
-	if (!buf) {
-		return rz_bin_dyldcache_free(bin);
+
+	cache_locsym_info_t *info = NULL;
+	char *strings = NULL;
+	cache_locsym_entry_t *entries = NULL;
+	struct MACH0_(nlist) *nlists = NULL;
+
+	ut64 info_size = sizeof(cache_locsym_info_t);
+	info = RZ_NEW0(cache_locsym_info_t);
+	if (!info) {
+		goto beach;
+	}
+	if (rz_buf_fread_at(cache_buf, hdr->localSymbolsOffset, (ut8 *)info, "6i", 1) != info_size) {
+		goto beach;
+	}
+
+	ut64 nlists_size = sizeof(struct MACH0_(nlist)) * info->nlistCount;
+	nlists = RZ_NEWS0(struct MACH0_(nlist), info->nlistCount);
+	if (!nlists) {
+		goto beach;
+	}
+	if (rz_buf_fread_at(cache_buf, hdr->localSymbolsOffset + info->nlistOffset, (ut8 *)nlists, "iccsl",
+		    info->nlistCount) != nlists_size) {
+		goto beach;
+	}
+
+	strings = malloc(info->stringsSize);
+	if (!strings) {
+		goto beach;
+	}
+	if (rz_buf_read_at(cache_buf, hdr->localSymbolsOffset + info->stringsOffset, (ut8 *)strings,
+		    info->stringsSize) != info->stringsSize) {
+		goto beach;
 	}
-	bin->b = rz_buf_new_with_bytes(NULL, 0);
-	if (!bin->b || !rz_buf_set_bytes(bin->b, buf, size)) {
-		return rz_bin_dyldcache_free(bin);
+
+	ut64 entries_size = sizeof(cache_locsym_entry_t) * info->entriesCount;
+	entries = RZ_NEWS0(cache_locsym_entry_t, info->entriesCount);
+	if (!entries) {
+		goto beach;
+	}
+	if (rz_buf_fread_at(cache_buf, hdr->localSymbolsOffset + info->entriesOffset, (ut8 *)entries, "3i",
+		    info->entriesCount) != entries_size) {
+		goto beach;
+	}
+
+	RzDyldLocSym *locsym = RZ_NEW0(RzDyldLocSym);
+	if (!locsym) {
+		goto beach;
+	}
+
+	locsym->nlists = nlists;
+	locsym->nlists_count = info->nlistCount;
+	locsym->strings = strings;
+	locsym->strings_size = info->stringsSize;
+	locsym->entries = entries;
+	locsym->entries_count = info->entriesCount;
+
+	free(info);
+
+	return locsym;
+
+beach:
+	free(info);
+	free(strings);
+	free(entries);
+	free(nlists);
+
+	eprintf("dyldcache: malformed local symbols metadata\n");
+	return NULL;
+}
+
+static void rz_dyld_locsym_free(RzDyldLocSym *locsym) {
+	if (!locsym) {
+		return;
+	}
+	RZ_FREE(locsym->strings);
+	RZ_FREE(locsym->entries);
+	RZ_FREE(locsym->nlists);
+	free(locsym);
+}
+
+RZ_API void rz_dyldcache_free(RzDyldCache *cache) {
+	if (!cache) {
+		return;
+	}
+
+	rz_list_free(cache->bins);
+	cache->bins = NULL;
+	rz_buf_free(cache->buf);
+	cache->buf = NULL;
+	if (cache->rebase_infos) {
+		int i;
+		for (i = 0; i < cache->rebase_infos->length; i++) {
+			rebase_info_free(cache->rebase_infos->entries[i].info);
+			cache->rebase_infos->entries[i].info = NULL;
+		}
+		RZ_FREE(cache->rebase_infos->entries);
+		RZ_FREE(cache->rebase_infos);
+	}
+	RZ_FREE(cache->hdr);
+	RZ_FREE(cache->maps);
+	RZ_FREE(cache->accel);
+	rz_dyld_locsym_free(cache->locsym);
+	RZ_FREE(cache);
```

Comment:
Here also. Please note, since we free `cache` here, previous `cache->accel = NULL` aren't necessary too.